File: algoritma_priority.py

processes = ['P1', 'P2', 'P3', 'P4']
burst_time = [10, 1, 2, 1]
priority = [3, 1, 4, 2]

def priority_scheduling(processes, burst_time, priority):
    n = len(processes)
    # Gabungkan dan urutkan berdasarkan priority
    combined = list(zip(priority, burst_time, processes))
    combined.sort()

    waiting_time = [0] * n
    turnaround_time = [0] * n


    for i in range(1, n):
        waiting_time[i] = combined[i - 1][1] + waiting_time[i - 1]
    

    for i in range(n):
        turnaround_time[i] = waiting_time[i] + combined[i][1]

    average_waiting_time = sum(waiting_time) / n
    average_turnaround_time = sum(turnaround_time) / n


    print("Langkah-langkah perhitungan waktu tunggu:")
    for i in range(n):
        if i == 0:
            print(f"Waktu tunggu {combined[i][2]} = 0 ms (proses pertama)")
        else:
            print(f"Waktu tunggu {combined[i][2]} = Waktu burst {combined[i - 1][2]} + Waktu tunggu {combined[i - 1][2]} = "
                  f"{combined[i - 1][1]} + {waiting_time[i - 1]} = {waiting_time[i]} ms")
    

    print("\nProses | Burst Time | Priority | Waiting Time | Turnaround Time")
    for i in range(n):
        print(f"{combined[i][2]:<7} | {combined[i][1]:<10} | {combined[i][0]:<8} | {waiting_time[i]:<12} | {turnaround_time[i]:<15}")
    
    print(f"\nAverage Waiting Time: {average_waiting_time:.2f} ms")
    print(f"Menghitung Average Waiting Time: (Total Waiting Time / Jumlah Proses) = ({sum(waiting_time)} / {n}) = {average_waiting_time:.2f} ms")

    print(f"\nAverage Turnaround Time: {average_turnaround_time:.2f} ms")
    print(f"Menghitung Average Turnaround Time: (Total Turnaround Time / Jumlah Proses) = ({sum(turnaround_time)} / {n}) = {average_turnaround_time:.2f} ms")


priority_scheduling(processes, burst_time, priority)

File: algoritma_priority.cpp
#include <iostream>
#include <algorithm>
#include <iomanip>
using namespace std;

struct Process {
    string name;
    int burst_time;
    int priority;
};


bool compare(Process a, Process b) {
    return a.priority < b.priority;
}


void priority_scheduling(Process processes[], int n) {
    sort(processes, processes + n, compare);
    
    int waiting_time[n];
    waiting_time[0] = 0; 


    for (int i = 1; i < n; i++) {
        waiting_time[i] = processes[i - 1].burst_time + waiting_time[i - 1];
    }


    int turnaround_time[n];
    for (int i = 0; i < n; i++) {
        turnaround_time[i] = waiting_time[i] + processes[i].burst_time;
    }

    float total_waiting_time = 0;
    for (int i = 0; i < n; i++) {
        total_waiting_time += waiting_time[i];
    }
    
    float average_waiting_time = total_waiting_time / n;
    float total_turnaround_time = 0;
    for (int i = 0; i < n; i++) {
        total_turnaround_time += turnaround_time[i];
    }
    float average_turnaround_time = total_turnaround_time / n;


    cout << "Langkah-langkah perhitungan waktu tunggu:\n";
    for (int i = 0; i < n; i++) {
        if (i == 0) {
            cout << "Waktu tunggu " << processes[i].name << " = 0 ms (proses pertama)\n";
        } else {
            cout << "Waktu tunggu " << processes[i].name << " = Waktu burst " << processes[i - 1].name 
                 << " + Waktu tunggu " << processes[i - 1].name << " = "
                 << processes[i - 1].burst_time << " + " << waiting_time[i - 1] << " = "
                 << waiting_time[i] << " ms\n";
        }
    }

    cout << "\nProses | Burst Time | Priority | Waiting Time | Turnaround Time\n";
    for (int i = 0; i < n; i++) {
        cout << setw(6) << processes[i].name << " | " 
             << setw(10) << processes[i].burst_time << " | " 
             << setw(8) << processes[i].priority << " | " 
             << setw(12) << waiting_time[i] << " | " 
             << setw(15) << turnaround_time[i] << endl;
    }
    
    cout << "\nAverage Waiting Time: " << average_waiting_time << " ms" << endl;
    cout << "Menghitung Average Waiting Time: (Total Waiting Time / Jumlah Proses) = ("
         << total_waiting_time << " / " << n << ") = " << average_waiting_time << " ms\n";
    
    cout << "\nAverage Turnaround Time: " << average_turnaround_time << " ms" << endl;
    cout << "Menghitung Average Turnaround Time: (Total Turnaround Time / Jumlah Proses) = ("
         << total_turnaround_time << " / " << n << ") = " << average_turnaround_time << " ms\n";
}

int main() {
    Process processes[] = {{"P1", 10, 3}, {"P2", 1, 1}, {"P3", 2, 4}, {"P4", 1, 2}};
    int n = sizeof(processes) / sizeof(processes[0]);
    priority_scheduling(processes, n);
    return 0;
}
